### PR TITLE
Fix use of ad cache

### PR DESCRIPTION
**Pull request by gammazero**
Ads were not being cached correctly due to inverted logic about when to cache.

Delete ad from datastore as soon as it is read into cache.

### CODE REVIEW

**Review by willscott — APPROVED** [review]

**Review by MarcoPolo — COMMENTED** [review]
I like the general idea of carrying the ad state along and passing it in explicitly. It was a refactor I had in mind with the ingester change.

**Comment by MarcoPolo** [comment]
Let's fix the race before merging please.

File: internal/ingest/linksystem.go [hunk 1]
```diff
@@ -195,10 +195,18 @@ func (ing *Ingester) ingestAd(publisher peer.ID, adCid cid.Cid) error {
...
 
-	ad, err := ing.loadAd(adCid, true)
+	ad, err := ing.loadAd(adCid)
 	if err != nil {
 		return fmt.Errorf("failed to load ad: %v", err)
 	}
+	// Cleanup ad cache now that ad is being processed. Functions deeper into
+	// the callstack get the ad as a parameter and not from the cache.
+	ing.adCacheMutex.Lock()
+	delete(ing.adCache, adCid)
```
[the remark this answers is not in the document]
Comment (gammazero) [reply internal/ingest/linksystem.go:205]: OK, we can move it when those PRs merge.

**Review by MarcoPolo — COMMENTED** [review]
I don't think this PR fix the actual race condition that happens when two works are processing the same ad with resync=true.

Say you have 2 workers (w1 & w2) and two sync events (s1 & s2):
```
s1: loadAd(adA) // inRunIngestStep; Deletes adA from datastore
s2: loadAd(adA) // Works because in cache resync=true.
w1: ingestAd(adA) // Finishes and deletes from the cache
w2: ingestAd(adA) // fails because cannot load adA from cache)
```


What if we don't even store the ad in the datastore and instead have three primitives:

```
1. storeAd(adCid, ad) (puts ad in ad map and increments counter by 1)
2. takeAd(adCid) -> ad, bool (decrement counter by 1 & returns ad if exists)
3. borrowAd(adCid) -> ad, bool (increment counter by 1 & returns ad if exists)
```

On some datastructure of the form:
```go
type adMapRC struct {
  counter sync.Map[adCid]uint32 (use atomic Add helpers for this counter)
  adMap sync.Map[adCid]ad
}
```

This is a reference counter for ads. When we load an ad during a sync, we increment the reference counter, and when we use that ad in an ingest worker we decrement the counter. When the counter is 0 we delete the ad from the map.

File: internal/ingest/linksystem.go [hunk 2]
```diff
@@ -195,10 +195,18 @@ func (ing *Ingester) ingestAd(publisher peer.ID, adCid cid.Cid) error {
 
 	log.Info("Processing advertisement")
 
-	ad, err := ing.loadAd(adCid, true)
+	ad, err := ing.loadAd(adCid)
```
[the remark this answers is not in the document]
Comment (gammazero) [reply internal/ingest/linksystem.go]: Done.  Got rid of ad cache completely, and now pass add as parameter after loading it once.

File: internal/ingest/linksystem.go [hunk 3]
```diff
@@ -406,39 +406,26 @@ func (ing *Ingester) indexContentBlock(adCid cid.Cid, pubID peer.ID, nentries ip
...
-			ing.adCache[adCid] = ad
-			ing.adCacheMutex.Unlock()
-		}
-	}()
-
 	if ok {
 		return ad, nil
 	}
 
 	// Getting the advertisement for the entries so we know
```
[the remark this answers is not in the document]
Comment (gammazero) [reply internal/ingest/linksystem.go:397]: nope.

File: internal/ingest/ingest.go [hunk 4]
```diff
@@ -290,12 +290,8 @@ func (ing *Ingester) Sync(ctx context.Context, peerID peer.ID, peerAddr multiadd
 		// earliest ad we've just synced to the latest.
 		if resync && len(seenAdCids) > 0 {
 			ing.markAdChainUnprocessed(seenAdCids)
```
[the remark this answers is not in the document]
Comment (gammazero) [reply internal/ingest/ingest.go:292]: Done

**Comment by gammazero** [comment]
Could the two-workers/same-ad problem be solved by storing the ad in the datastore using a key prefixed by the publisher peer ID.

**Comment by MarcoPolo** [comment]
I don't think that would solve it. Consider this flow that could randomly happen:

1. sync & resync happening back to back for the same publisher.
2. legs sync writes to the datastore. The resync sees it's in the datastore so no need to fetch.
3. Emit two syncFinished events back to back.
4. First one from the normal sync happens, and loads the ad in runIngestStep. Then sends it to the worker.
5. The second one fails to find the ad when it calls `loadAd` in runIngestStep because it was just deleted in 4.

I think there needs to be some way of knowing if there's another thing that will need this ad before we delete. Hence the reference counter.

**Comment by gammazero** [comment]
> 1. sync & resync happening back to back for the same publisher.

legs serializes syncs with the same publisher

**Comment by MarcoPolo** [comment]
> legs serializes syncs with the same publisher
But this could happen even if the sync happens back to back (not at the same time)

**Comment by gammazero** [comment]
Agreed to solve two-workers/same-ad problem in a separate PR, since it is an edge case.

**Review by MarcoPolo — APPROVED** [review]
This still has the race condition, but I don't think it's any worse than it is in main.

And this cleans up the adCache business. so worth merging.

**Pull request merged**